src/translate.py

```diff
@@ -3717,20 +3701,147 @@ def symbol_type(self, sym_name: str) -> Type:
         lines = []
-        for name, type in self.symbol_type_map.items():
-            if self.typemap and name in self.typemap.var_types:
-                continue
-            is_static = name in self.asm_data.values
-            order = (is_static, name)
-            prefix = ""
-            if is_static:
-                prefix = "static "
-            elif not type.is_function():
-                prefix = "extern "
-
-            lines.append((order, f"{prefix}{type.to_decl(name, fmt)};\n"))
+        processed_names: Set[str] = set()
+        while True:
+            names = self.symbol_type_map.keys() - processed_names
+            if not names:
+                break
+            for name in names:
+                processed_names.add(name)
+                type = self.symbol_type_map[name]
+
+                # Is the label defined in this unit (in the active AsmData file(s))
+                is_in_file = (
+                    name in self.asm_data.values or name in self.local_functions
+                )
+                # Is the label externally visible (mentioned in the context file)
+                is_global = self.typemap and name in self.typemap.var_types
+                if not is_in_file and is_global:
+                    # Skip externally-declared symbols that are defined in other files
+                    continue
+                if is_in_file and is_global and type.is_function():
+                    # Skip externally-declared functions that are defined here
+                    continue
+
+                sort_order = (not type.is_function(), is_global, is_in_file, name)
+                qualifier = ""
+                initializer = ""
+                comments = []
+
+                # Determine type qualifier: static, extern, or neither
+                if is_in_file and is_global:
+                    qualifier = ""
+                elif is_in_file:
+                    qualifier = "static"
+                else:
+                    qualifier = "extern"
+
+                if type.is_function():
+                    comments.append(qualifier)
+                    qualifier = ""
+
+                # Try to convert the data from .data/.rodata into an initializer
+                entry = self.asm_data.values.get(name)
+                if entry is not None:
+                    if entry.is_readonly:
+                        comments.append("const")
+                    data_size = entry.size_bytes()
+                    type_size = type.get_size_bits() // 8
+                    if data_size != type_size:
+                        # TODO: Use heuristic to detect padding
+                        if data_size // type_size > 1:
+                            comments.append(f"array: [{data_size // type_size}]")
+                        if data_size % type_size != 0:
+                            comments.append(f"extra bytes: {data_size % type_size}")
+                    elif entry.is_readonly and type.is_float():
```

Comment:
maybe move this above the other check

Reply:
Oh, this was intentional, to not skip float arrays?
(It'd be better to have a way of marking which constants need to be emitted, but TODO?)